src/coreclr/jit/gentree.cpp

```diff
@@ -3835,10 +3835,17 @@ bool Compiler::gtIsLikelyRegVar(GenTree* tree)
 //
 bool Compiler::gtCanSwapOrder(GenTree* firstNode, GenTree* secondNode)
 {
-    // Relative of order of global / side effects can't be swapped.
-
     bool canSwap = true;
 
+    // Don't swap "CONST_HDL op CNS"
+    if (firstNode->IsIntegralConst() && secondNode->IsIntegralConst() && varTypeIsGC(firstNode) &&
+        !varTypeIsGC(secondNode))
+    {
+        canSwap = false;
+    }
```

Comment:
IIUC there is already some code in `gtMarkAddrMode` that is meant to handle the case:
https://github.com/dotnet/runtime/blob/948142214f24c0b2db485e9246b73f18be43cf49/src/coreclr/jit/codegencommon.cpp#L1129-L1131

Does it just need to use `varTypeIsGC` instead or what was the failure here?
I guess we might see the situation even without the JIT swapping the addition on its own.

Reply:
Yes I tried that first but then decided to go with my change - a sort of canonization that "harmless" constants are always on the right, we can do both 🤷‍♂️ but so far it doesn't repro any more

Reply:
Ah okay, sounds fine to me. Was it just a CQ issue or was it an actual codegen/assertion failure?

Reply:
> Ah okay, sounds fine to me. Was it just a CQ issue or was it an actual codegen/assertion failure?

It was a debug-only assert in `gtMarkAddrMode` - there is a weird logic I am tempted to remove that basically mirrors `gtMarkAddrMode` to validate it - https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/gentree.cpp#L4143-L4254
it's also eating a lot of JIT TP